### PR TITLE
Add ExecutionEngine error messages to improve SC debugging feedback

### PR DESCRIPTION
**Pull request by ixje**
**What current issue(s) does this address, or what feature is it adding?**
A continuation of #294 , specifically point 4 (the final one)
>add additional error messages to ExecutionEngine.
rationale; there's a lot of error feedback we can simply give that we're not doing leading to guess work or a lot of stepping through the VM internals.

**How did you solve this problem?**
As discussed here
> How about the following for 4;
> all interesting spots for adding additional error information seem to end with:
>                     
> self._VMState |= VMState.FAULT
> return
> 
> what if we replace that with an error reporting function e.g.
> def VM_FAULT_and_report(id, *args):
>     self._VMState |= VMState.FAULT
> 
>     # additional error handling/message providing based on id
>     ...
> 
> That way we can keep further handling all in 1 spot and keep ExecutionEngine as bare as possible

> March 8, 2018
> localhuman - Last Thursday at 1:55 PM
> I think its a good approach

**How did you make sure your solution works?**
PoC, no tests yet.

**Are there any special changes in the code that we should be aware of?**
no

**Please check the following, if applicable:**

- [X] Did you add any tests?
- [X] Did you run `make lint`?
- [X] Did you run `make test`?
- [X] Are you making a PR to a feature branch or development rather than master?
- [X] Did you add an entry to `CHANGELOG.rst`? (if not, please do)

### CODE REVIEW

**Comment by localhuman** [comment]
I think it looks good but I worry about maintaining the integers for the error reports and keeping them matched up between the execution and the reporting. 

 So instead of doing `self.VM_FAULT_and_report(2)` it would be `self.VM_FAULT_and_report( VMFAULT.INVALID_JUMP)`

Does that make sense?

**Comment by metachris** [comment]
Agree, makes a lot of sense to use constants for this!

**Comment by ixje** [comment]
@metachris Note that I didn't use the testing approach you gave me, because I couldn't get `capsys` from the `pytest` package to work that you used for capturing the `stdoutput/err`. I'm guessing it requires a different test runner than what we use in `neo-python`. Either way, my current approach doesn't require a new package so I guess it should suffice.

File: neo/VM/ExecutionEngine.py [hunk 1]
```diff
@@ -669,8 +650,7 @@ def ExecuteOp(self, opcode, context):
                 size = estack.Pop().GetBigInteger()
 
                 if size < 0 or size > estack.Count:
-                    self._VMState |= VMState.FAULT
-                    return
+                    return self.VM_FAULT_and_report(18)
```
Comment (localhuman) [inline neo/VM/ExecutionEngine.py]: should `18` be a `VMFault`?

File: neo/VM/ExecutionEngine.py [hunk 2]
```diff
@@ -246,8 +238,7 @@ def ExecuteOp(self, opcode, context):
                 n = estack.Pop().GetBigInteger()
 
                 if n < 0:
-                    self._VMState |= VMState.FAULT
-                    return
+                    return self.VM_FAULT_and_report(VMFault.UNKNOWN3)
```
Comment (localhuman) [inline neo/VM/ExecutionEngine.py]: Will `UNKNOWN3-6` be replaced with something?

File: neo/VM/ExecutionEngine.py (hunk 1 above)
Comment (ixje) [reply neo/VM/ExecutionEngine.py]: oops missed

File: neo/VM/ExecutionEngine.py (hunk 2 above)
Comment (ixje) [reply neo/VM/ExecutionEngine.py]: For now not, unless you have a clear idea what a good fault statement would be for those. I decided to keep them `generic unknown` until with some more experience I figure out a good name for it.

**Comment by localhuman** [comment]
Nice work!

**Pull request merged**